File: goldencheck/llm/merger.py
"""Merge LLM response into existing findings list."""
from __future__ import annotations
import logging
import re
from dataclasses import replace
from goldencheck.llm.prompts import LLMResponse
from goldencheck.models.finding import Finding, Severity

logger = logging.getLogger(__name__)

SEVERITY_MAP = {"error": Severity.ERROR, "warning": Severity.WARNING, "info": Severity.INFO}

# Required keywords per check — if LLM message lacks them, append a suffix
_REQUIRED_KEYWORDS: dict[str, list[str]] = {
    "cross_column": ["mismatch", "inconsistent", "doesn't match"],
    "invalid_values": ["invalid"],
}


def _ensure_keywords(check: str, message: str) -> str:
    """Ensure the message contains at least one required keyword for the check."""
    keywords = _REQUIRED_KEYWORDS.get(check, [])
    if not keywords:
        return message
    msg_lower = message.lower()
    if any(kw in msg_lower for kw in keywords):
        return message
    suffix = {
        "cross_column": " [cross-column mismatch detected]",
        "invalid_values": " [invalid values detected]",
    }
    return message + suffix.get(check, "")


def _strip_suppression_suffix(message: str) -> str:
    return re.sub(r'\s*\(suppressed:.*?\)\s*$', '', message)


def merge_llm_findings(
    findings: list[Finding],
    response: LLMResponse | None,
) -> list[Finding]:
    """Merge LLM response into findings. Returns a new list (never mutates originals)."""
    if response is None:
        return list(findings)

    result = list(findings)

    # Build lookup index: (column, check) -> index in result
    index = {}
    for i, f in enumerate(result):
        index[(f.column, f.check)] = i

    # Process per-column assessments
    for col_name, assessment in response.columns.items():
        # New issues
        for issue in assessment.issues:
            sev = SEVERITY_MAP.get(issue.severity.lower(), Severity.WARNING)
            result.append(Finding(
                severity=sev,
                column=col_name,
                check=issue.check,
                message=_ensure_keywords(issue.check, issue.message),
                sample_values=issue.affected_values,
                source="llm",
            ))

        # Upgrades (use dataclasses.replace to avoid mutation)
        for upgrade in assessment.upgrades:
            key = (col_name, upgrade.original_check)
            if key in index:
                old = result[index[key]]
                result[index[key]] = replace(
                    old,
                    severity=SEVERITY_MAP.get(upgrade.new_severity.lower(), old.severity),
                    message=f"{_strip_suppression_suffix(old.message)} [LLM: {upgrade.reason}]",
                    source="llm",
                )
            else:
                # Create as new issue
                result.append(Finding(
                    severity=SEVERITY_MAP.get(upgrade.new_severity.lower(), Severity.WARNING),
                    column=col_name,
                    check=upgrade.original_check,
                    message=upgrade.reason,
                    source="llm",
                ))

        # Downgrades (use dataclasses.replace to avoid mutation)
        for downgrade in assessment.downgrades:
            key = (col_name, downgrade.original_check)
            if key in index:
                old = result[index[key]]
                result[index[key]] = replace(
                    old,
                    severity=SEVERITY_MAP.get(downgrade.new_severity.lower(), old.severity),
                    message=f"{_strip_suppression_suffix(old.message)} [LLM: {downgrade.reason}]",
                    source="llm",
                )
            # else: silently ignore

    # Process relations
    for relation in response.relations:
        col_key = ",".join(sorted(relation.columns))
        result.append(Finding(
            severity=Severity.WARNING,
            column=col_key,
            check=relation.type,
            message=relation.reasoning,
            source="llm",
        ))

    return result

File: packages/goldencheck-js/src/core/llm/merger.ts
/**
 * Merge LLM response into existing findings list (immutable).
 * Port of goldencheck/llm/merger.py.
 * Edge-safe: no Node.js dependencies.
 */

import type { Finding } from "../types.js";
import { Severity, makeFinding, replaceFinding } from "../types.js";
import type { LLMResponse } from "./prompts.js";

const SEVERITY_MAP: Record<string, (typeof Severity)[keyof typeof Severity]> = {
  error: Severity.ERROR,
  warning: Severity.WARNING,
  info: Severity.INFO,
};

/**
 * Required keywords per check name.
 * If the LLM message lacks them, a suffix is appended.
 */
const REQUIRED_KEYWORDS: Record<string, string[]> = {
  cross_column: ["mismatch", "inconsistent", "doesn't match"],
  invalid_values: ["invalid"],
};

const KEYWORD_SUFFIXES: Record<string, string> = {
  cross_column: " [cross-column mismatch detected]",
  invalid_values: " [invalid values detected]",
};

/** Ensure the message contains at least one required keyword for the check. */
function ensureKeywords(check: string, message: string): string {
  const keywords = REQUIRED_KEYWORDS[check];
  if (!keywords) return message;
  const msgLower = message.toLowerCase();
  if (keywords.some((kw) => msgLower.includes(kw))) return message;
  return message + (KEYWORD_SUFFIXES[check] ?? "");
}

/** Strip existing "(suppressed: ...)" suffixes before merging. */
function stripSuppressionSuffix(message: string): string {
  return message.replace(/\s*\(suppressed:.*?\)\s*$/, "");
}

/**
 * Merge LLM response into findings. Returns a new list (never mutates originals).
 *
 * - New issues: appended with source="llm"
 * - Upgrades: matched by (column, check), severity updated, reason appended
 * - Downgrades: matched by (column, check), severity updated, reason appended
 * - Relations: appended as Finding(column="col_a,col_b", source="llm")
 */
export function mergeLlmFindings(
  findings: readonly Finding[],
  response: LLMResponse | null,
): Finding[] {
  if (response === null) return [...findings];

  const result = [...findings];

  // Build lookup index: (column, check) -> index in result
  const index = new Map<string, number>();
  for (let i = 0; i < result.length; i++) {
    const f = result[i]!;
    index.set(`${f.column}\0${f.check}`, i);
  }

  // Process per-column assessments
  for (const [colName, assessment] of Object.entries(response.columns)) {
    // New issues
    for (const issue of assessment.issues) {
      const sev = SEVERITY_MAP[issue.severity.toLowerCase()] ?? Severity.WARNING;
      result.push(
        makeFinding({
          severity: sev,
          column: colName,
          check: issue.check,
          message: ensureKeywords(issue.check, issue.message),
          sampleValues: issue.affected_values ?? [],
          source: "llm",
        }),
      );
    }

    // Upgrades (immutable via replaceFinding)
    for (const upgrade of assessment.upgrades) {
      const key = `${colName}\0${upgrade.original_check}`;
      const idx = index.get(key);
      if (idx !== undefined) {
        const old = result[idx]!;
        result[idx] = replaceFinding(old, {
          severity: SEVERITY_MAP[upgrade.new_severity.toLowerCase()] ?? old.severity,
          message: `${stripSuppressionSuffix(old.message)} [LLM: ${upgrade.reason}]`,
          source: "llm",
        });
      } else {
        // Create as new issue when original not found
        result.push(
          makeFinding({
            severity: SEVERITY_MAP[upgrade.new_severity.toLowerCase()] ?? Severity.WARNING,
            column: colName,
            check: upgrade.original_check,
            message: upgrade.reason,
            source: "llm",
          }),
        );
      }
    }

    // Downgrades (immutable via replaceFinding)
    for (const downgrade of assessment.downgrades) {
      const key = `${colName}\0${downgrade.original_check}`;
      const idx = index.get(key);
      if (idx !== undefined) {
        const old = result[idx]!;
        result[idx] = replaceFinding(old, {
          severity: SEVERITY_MAP[downgrade.new_severity.toLowerCase()] ?? old.severity,
          message: `${stripSuppressionSuffix(old.message)} [LLM: ${downgrade.reason}]`,
          source: "llm",
        });
      }
      // else: silently ignore downgrades for findings that don't exist
    }
  }

  // Process relations
  for (const relation of response.relations) {
    const colKey = [...relation.columns].sort().join(",");
    result.push(
      makeFinding({
        severity: Severity.WARNING,
        column: colKey,
        check: relation.type,
        message: relation.reasoning,
        source: "llm",
      }),
    );
  }

  return result;
}
